usr/lib/common/pkey_utils.c

```diff
@@ -1196,31 +1196,52 @@ CK_RV pkey_ec_sign(STDLL_TokData_t *tokdata, SESSION *session,
     curve_type = get_cpacf_curve_type(privkey->template);
     switch (curve_type) {
     case curve_p256:
+        if (pkey_attr->ulValueLen != sizeof(param.P256.priv) + sizeof(param.P256.vp)) {
+            TRACE_ERROR("Protected key has an invalid length of %ld bytes.\n",
+                        pkey_attr->ulValueLen);
+            ret = CKR_ATTRIBUTE_VALUE_INVALID;
+            goto done;
+        }
         if (hash_len > 32)
             hash_len = 32;
         off = 32 - hash_len;
         memcpy(param.P256.hash + off, hash, hash_len);
-        memcpy(param.P256.priv, pkey_attr->pValue, 32);
-        memcpy(param.P256.vp, (char *)pkey_attr->pValue + 32, 32);
+        memcpy(param.P256.priv, pkey_attr->pValue,
+               MIN(pkey_attr->ulValueLen - 32, sizeof(param.P256.priv)));
+        memcpy(param.P256.vp, (char *)pkey_attr->pValue + 32, sizeof(param.P256.vp));
```

Comment:
Since you have the explicit check above, you can omit the MIN here. 

However, instead of the hardcoded +/- 32, I would use `sizeof(xxxx)` like in the following:

```
memcpy(param.P256.priv, pkey_attr->pValue, pkey_attr->ulValueLen - sizeof(param.P256.vp), sizeof(param.P256.priv)));
memcpy(param.P256.vp, (char *)pkey_attr->pValue + sizeof(param.P256.priv), sizeof(param.P256.vp));
```

Similar for the `32` for the hash, can we use a `sizeof(something)` here ? Isn't `sizeof(param.P256.hash)` exactly what you need ?

Reply:
It's the vp that must be subtracted. I removed the MIN(...) for p256, p384, and p521 and made 1 single line then. Let me know if you want to have a line break before col 80 .... it's always some tradeoff between readability and format conventions.

Reply:
Yes please no lines linger than 80.

But you still have a +32 in the second memcpy:
`memcpy(param.P256.vp, (char *)pkey_attr->pValue + 32, sizeof(param.P256.vp));`
I guess that one should be
`memcpy(param.P256.vp, (char *)pkey_attr->pValue + sizeof(param.P256.priv), sizeof(param.P256.vp));`

Reply:
ok, done.